app/upgrades/v12/upgrade_test.go

```diff
@@ -0,0 +1,90 @@
+package v12_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+
+	"github.com/osmosis-labs/osmosis/v11/app/apptesting"
+)
+
+type UpgradeTestSuite struct {
+	apptesting.KeeperTestHelper
+}
+
+func (suite *UpgradeTestSuite) SetupTest() {
+	suite.Setup()
+}
+
+func TestKeeperTestSuite(t *testing.T) {
+	suite.Run(t, new(UpgradeTestSuite))
+}
+
+const dummyUpgradeHeight = 5
+
+// note that this test does not perfectly test state migration, as pre_upgrade function
+// which includes setting up tests by creating new pools would also create twap records
+// automatically with this binary. The goal of this test is to test that the upgrade handler
+// does not panic upon upgrade.
+// Detailed state migration tests are placed within the twap keeper.
+func (suite *UpgradeTestSuite) TestPoolMigration() {
+	testCases := []struct {
+		name         string
+		pre_upgrade  func() uint64
+		upgrade      func()
+		post_upgrade func(uint64)
+	}{
+		{
+			"Test that the upgrade succeeds",
+			func() uint64 {
+				poolId := suite.PrepareBalancerPool()
+				poolDenoms, err := suite.App.GAMMKeeper.GetPoolDenoms(suite.Ctx, poolId)
+				suite.Require().NoError(err)
+
+				_, err = suite.App.TwapKeeper.GetBeginBlockAccumulatorRecord(suite.Ctx, poolId, poolDenoms[0], poolDenoms[1])
+				suite.Require().NoError(err)
+				return poolId
+			},
+			func() {
+				suite.Ctx = suite.Ctx.WithBlockHeight(dummyUpgradeHeight - 1)
+				plan := upgradetypes.Plan{Name: "v12", Height: dummyUpgradeHeight}
+				err := suite.App.UpgradeKeeper.ScheduleUpgrade(suite.Ctx, plan)
+				suite.Require().NoError(err)
+				plan, exists := suite.App.UpgradeKeeper.GetUpgradePlan(suite.Ctx)
+				suite.Require().True(exists)
+
+				suite.Ctx = suite.Ctx.WithBlockHeight(dummyUpgradeHeight)
+				suite.Require().NotPanics(func() {
+					beginBlockRequest := abci.RequestBeginBlock{}
+					suite.App.BeginBlocker(suite.Ctx, beginBlockRequest)
+				})
+			},
+			func(poolId uint64) {
+				poolDenoms, err := suite.App.GAMMKeeper.GetPoolDenoms(suite.Ctx, poolId)
+				suite.Require().NoError(err)
+
+				_, err = suite.App.TwapKeeper.GetBeginBlockAccumulatorRecord(suite.Ctx, poolId, poolDenoms[0], poolDenoms[1])
+				suite.Require().NoError(err)
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("Case %s", tc.name), func() {
+			suite.SetupTest() // reset
+
+			// creting pools before upgrade
+			poolId := tc.pre_upgrade()
+
+			// run upgrade
+			tc.upgrade()
```

Comment:
Is it possible to avoid using callbacks? I think it would make sense if there were multiple cases with different `pre_upgrade`, `upgrade` or `post_upgrade` logic but here it seems that we have only 1

Also, are we expecting more test cases? Does it make sense for this test to not be table-driven?

Reply:
I think we should come back in a followup, and cleanup the entire go upgrade test framework. This is copied from another upgrade test, we should make a unified setup for all of them / decide what we generally want.

I think we should capture this into an issue